build.rs

```diff
@@ -12,14 +12,23 @@
 // limitations under the License.
 
 use regex::Regex;
+use std::env;
+use std::fs::read_dir;
 use std::fs::File;
-use std::fs::{read_dir, remove_file};
 use std::io::{Read, Write};
 use std::process::Command;
 
 fn main() {
+    // This build script creates files in the `src` directory. Since that is
+    // outside Cargo's OUT_DIR it will cause an error when this crate is used
+    // as a dependency. Therefore, the user must opt-in to regenerating the
+    // Rust files.
+    if env::var_os("CARGO_FEATURE_REGENERATE").is_none() {
+        println!("cargo:rerun-if-changed=build.rs");
```

Comment:
Shouldn't the changes inside `proto` directory trigger rebuild too?

Reply:
I think only if we are regenerating the proto files

Reply:
Changes to proto files only make sense when code is generated.

Reply:
aha, can we regenerate them all the time?

Reply:
This is not a good idea. If we use kvproto as a dependency, all the `.proto`s are treated as "changed" locally so a regeneration will get triggered if we do so.

Reply:
In my case, when building TiKV, I get this message:

```
--- stdout
cargo:rerun-if-changed=proto/errorpb.proto
cargo:rerun-if-changed=proto/kvrpcpb.proto
cargo:rerun-if-changed=proto/raft_serverpb.proto
cargo:rerun-if-changed=proto/coprocessor.proto
cargo:rerun-if-changed=proto/pdpb.proto
cargo:rerun-if-changed=proto/import_sstpb.proto
cargo:rerun-if-changed=proto/metapb.proto
cargo:rerun-if-changed=proto/raft_cmdpb.proto
cargo:rerun-if-changed=proto/debugpb.proto
cargo:rerun-if-changed=proto/tikvpb.proto
cargo:rerun-if-changed=proto/import_kvpb.proto
```

where I don't expect one.